contracts/epochs/FlowEpoch.cdc

```diff
@@ -432,6 +559,97 @@ access(all) contract FlowEpoch {
+            } else {
+                /// force reset the QC and DKG
+                FlowEpoch.borrowClusterQCAdmin().forceStopVoting()
+                FlowEpoch.borrowDKGAdmin().forceEndDKG()
+            }
+
+            /// Create the recovery epoch metadata, this struct is used to hold 
+            /// EpochRecover service event data.
+            let recoverEpochMetadata = FlowEpoch.RecoverEpochMetadata(
+                counter: FlowEpoch.proposedEpochCounter(),
+                nodeIDs: nodeIDs,
+                firstView: startView,
+                finalView: endView,
+                collectorClusters: collectorClusters,
+                randomSource: randomSource,
+                dkgPhase1FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + FlowEpoch.configurableMetadata.numViewsInDKGPhase - 1 as UInt64,
+                dkgPhase2FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + (2 as UInt64 * FlowEpoch.configurableMetadata.numViewsInDKGPhase) - 1 as UInt64,
+                dkgPhase3FinalView: startView + FlowEpoch.configurableMetadata.numViewsInStakingAuction + (3 as UInt64 * FlowEpoch.configurableMetadata.numViewsInDKGPhase) - 1 as UInt64,
+                targetDuration: targetDuration,
+                targetEndTime: targetEndTime,
+                clusterQCVoteData: clusterQCVoteData,
+                dkgPubKeys: dkgPubKeys,
+            )
+
+            /// Save the recovery epoch metadata to storage, this will be emitted in the EpochRecover service event
+            /// during the next heart beat interval.
+            FlowEpoch.account.storage.save(recoverEpochMetadata, to: /storage/recoverEpochMetadataStoragePath)
+
+            /// Create new EpochMetadata for the recovery epoch with the new values
+            let newEpochMetadata = EpochMetadata(
+                /// When the network enters EFM the epoch counter will not be incremented 
+                /// so it's safe to use current_epoch_counter + 1 
+                counter: FlowEpoch.proposedEpochCounter(),
+                seed: randomSource,
+                startView: startView,
+                endView: endView,
+                stakingEndView: stakingEndView,
+                // The following fields will be overwritten in `calculateAndSetRewards` below
+                totalRewards: UFix64(0.0),
+                collectorClusters: [],
+                clusterQCs: [],
+                dkgKeys: dkgPubKeys)
+
+            /// Save the new epoch meta data, it will be referenced as 
+            /// the epoch progresses through each phase.
+            FlowEpoch.saveEpochMetadata(newEpochMetadata)
+
+            /// Calculate rewards for the current epoch
+            /// and set the payout for the next epoch
+            FlowEpoch.calculateAndSetRewards()
+            /// Start a new Epoch, which increments the current epoch counter
+            FlowEpoch.startNewEpoch()
```

Comment:
It's possible that the `EpochRecover` will be rejected by the Protocol State, in which case we would need to send a new `recoverEpoch` transaction.

We should make sure this function is decoupled from reward payout so we can, if necessary, re-try this process without double-paying rewards.